client-state-machine.go

```diff
@@ -818,44 +860,67 @@ func (state clientStateWaitCV) Next(hr handshakeMessageReader) (HandshakeState,
 	hcv := state.handshakeHash.Sum(nil)
 	logf(logTypeHandshake, "Handshake Hash to be verified: [%d] %x", len(hcv), hcv)
 
-	serverPublicKey := state.serverCertificate.CertificateList[0].CertData.PublicKey
+	certs := make([][]byte, len(state.serverCertificate.CertificateList))
+	for i, certEntry := range state.serverCertificate.CertificateList {
+		certs[i] = certEntry.CertData
+	}
+
+	var err error
+	var serverPublicKey crypto.PublicKey
+	var eeCert *x509.Certificate
+	if !state.Params.UsingRawPublicKeys {
+		eeCert, err = x509.ParseCertificate(certs[0])
+		if err != nil {
+			logf(logTypeHandshake, "[ClientStateWaitCV] Unable to parse client cert: %v", err)
+		}
+
+		serverPublicKey = eeCert.PublicKey
+	} else {
+		serverPublicKey, err = unmarshalSigningKey(certs[0])
+		if err != nil {
+			logf(logTypeHandshake, "[ClientStateWaitCV] Unable to parse raw public key: %v", err)
+		}
+	}
```

Comment:
BUG: You don't fail the handshake here.

Reply:
Good point.  Same actually applies for the certificate parsing a few lines above.